x/ccv/consumer/keeper/distribution_v2.go

```diff
@@ -0,0 +1,251 @@
+}
+
+func (k Keeper) DistributeToProviderValidatorSetV2(ctx sdk.Context) (err error) {
+	ltbh, err := k.GetLastTransmissionBlockHeight(ctx)
+	if err != nil {
+		return err
+	}
+	bpdt := k.GetBlocksPerDistributionTransmission(ctx)
+	curHeight := ctx.BlockHeight()
+
+	if (curHeight - ltbh.Height) < bpdt {
+		// not enough blocks have passed for  a transmission to occur
+		return nil
+	}
+
+	defer func() {
+		if err != nil {
+			newLtbh := types.LastTransmissionBlockHeight{
+				Height: ctx.BlockHeight(),
+			}
+			err = k.SetLastTransmissionBlockHeight(ctx, newLtbh)
+		}
+	}()
+
+	addresses := make([][]byte, 0)
+	weights := make([]sdk.Int, 0)
+	totalWeights := sdk.ZeroInt()
+	k.IterateValidatorHoldingPools(ctx, func(valAddr []byte, weight sdk.Int) bool {
+		addresses = append(addresses, valAddr)
+		weights = append(weights, weight)
+		totalWeights = totalWeights.Add(weight)
+
+		k.DeleteValidatorHoldingPool(ctx, valAddr) // clear the holding pool
+		return false
+	})
+
+	providerRewardStagingAddress := k.authKeeper.GetModuleAccount(ctx, types.ProviderRewardStagingName).GetAddress()
+	tokensInStagingAddress := k.bankKeeper.GetAllBalances(ctx, providerRewardStagingAddress)
+	providerPoolWeights := ccv.ProviderPoolWeights{
+		Addresses:   addresses,
+		Weights:     weights,
+		TotalWeight: totalWeights,
+		Tokens:      tokensInStagingAddress,
+	}
+
+	// transfer tokens from reward staging address to toSendToProviderTokens address
+	if err = k.bankKeeper.SendCoinsFromModuleToModule(ctx, types.ProviderRewardStagingName,
+		types.ConsumerToSendToProviderName, tokensInStagingAddress); err != nil {
+		return err
+	}
+
+	// transfer tokens from toSendToProviderTokens address to provider via IBC
+	transmissionChannel := k.GetDistributionTransmissionChannel(ctx)
+	if len(transmissionChannel) != 0 {
+		// empty out the toSendToProviderTokens address
+		tstProviderAddr := k.authKeeper.GetModuleAccount(ctx,
+			types.ConsumerToSendToProviderName).GetAddress()
+		tstProviderTokens := k.bankKeeper.GetAllBalances(ctx, tstProviderAddr)
+		providerAddr := k.GetProviderFeePoolAddrStr(ctx)
+		timeoutHeight := clienttypes.ZeroHeight()
+		timeoutTimestamp := uint64(ctx.BlockTime().Add(TransferTimeDelay).UnixNano())
+		for _, token := range tstProviderTokens {
+			if err = k.ibcTransferKeeper.SendTransfer(ctx,
+				transfertypes.PortID,
+				transmissionChannel,
+				token,
+				tstProviderAddr,
+				providerAddr,
+				timeoutHeight,
+				timeoutTimestamp,
+			); err != nil {
+				return err
+			}
+		}
+	}
+
+	// if not, append provider pool weights to pending provider pool weights
+	channelID, ok := k.GetProviderChannel(ctx)
+	if !ok {
+		k.AppendPendingProviderPoolWeights(ctx, providerPoolWeights)
+		return
+	}
+
+	// send packet over IBC
+	return utils.SendIBCPacket(
+		ctx,
+		k.scopedKeeper,
+		k.channelKeeper,
+		channelID,    // source channel id
+		types.PortID, // source port id
+		providerPoolWeights.GetBytes(),
+	)
+}
```

Comment:
what happened if providerPoolWeights greater or smaller than tokens in SendTransfer?
we have the check: "if balances.Len() == 0 || ppw.Tokens.IsAnyGT(balances) {" indeed
but what if packet message delayed and cumulate balance is gt providerPoolWeights then providerPoolWeights is gt the actual "SendTransfer" token amount?

Reply:
I think we need some "transferID" check on provider side between SendTransfer token amount and providerPoolWeights tokens

Reply:
> what happened if providerPoolWeights greater or smaller than tokens in SendTransfer? we have the check: "if balances.Len() == 0 || ppw.Tokens.IsAnyGT(balances) {" indeed but what if packet message delayed and cumulate balance is gt providerPoolWeights then providerPoolWeights is gt the actual "SendTransfer" token amount?

it is not possible that `the providerPoolWeights gt the actual "SendTransfer" token amount`, it is promised by the consumer side logic. Besides, we can get any like 'transferID' from IBCTransfer method at the consumer side, and the only thing we can see at the provider side is that the balance added when an IBCTransfer packet is delivered.

Reply:
we can't assumed something will be promised by consumer side logic